projects/output-and-error-handling/main.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+)
+
+func sendRequest(url string) *http.Response {
+	resp, err := http.Get(url)
+	if err != nil {
+		fmt.Println("We can't get the weather!")
+		fmt.Fprintln(os.Stderr, "Failed to initialize.!", err)
+		os.Exit(2)
+	}
+	if resp == nil {
+		fmt.Fprintln(os.Stderr, "The server did not reply anything, which here is considered an error!")
+		os.Exit(52)
+	}
+	return resp
+}
+
+func handleStatusCode(resp *http.Response, sleepTime *time.Duration) {
+	defer resp.Body.Close()
+	if resp.StatusCode == 200 {
+		scanner := bufio.NewScanner(resp.Body)
+		for i := 0; scanner.Scan(); i++ {
+			fmt.Println(scanner.Text())
+		}
+		if err := scanner.Err(); err != nil {
+			panic(err)
+		}
+		os.Exit(0)
+	} else if resp.StatusCode == 429 {
+		// If I receive 429 but can't find ["Retry-After"] so I determine 1s to sleep. It act same as a while
+		retryAfter := resp.Header.Get("Retry-After")
+		if retryAfter == "" || retryAfter == "a while" {
+			// I added with the previous sleep times to prevent sleeping for too long and more than 5s give up.
+			*sleepTime += time.Second
+		} else {
+
+			*sleepTime = parseRetryAfterNumber(retryAfter)
+		}
+		if *sleepTime > 5*time.Second {
+			fmt.Println("We can't get the weather!")
+			fmt.Fprintln(os.Stderr, "Sleep time is more than 5 second so HTTP page not retrieved. Returned another error with the HTTP error code being 400 or above.")
+			os.Exit(22)
+		} else if *sleepTime > time.Second {
+			fmt.Println("Things may be a bit slow because we're doing a retry!")
+		}
+		time.Sleep(*sleepTime)
+	} else {
+		fmt.Fprintln(os.Stderr, "Failed to initialize!")
+		os.Exit(2)
+	}
+}
+
+func parseRetryAfterNumber(retryAfter string) time.Duration {
+	if sleepTime, err := time.ParseDuration(retryAfter + "s"); err != nil {
+		return parseRetryAfterString(retryAfter)
+	} else {
+		return sleepTime
+	}
+}
+
+func parseRetryAfterString(retryAfter string) time.Duration {
+	sleepTime := time.Second
+	if timeStamp, err := time.Parse(http.TimeFormat, retryAfter); err == nil {
+		sleepTime = time.Until(timeStamp)
+	} else {
+		fmt.Fprintln(os.Stderr, "There was an error in the value of Retry-After", err)
+	}
+	return sleepTime
+}
```

Comment:
It's really interesting that these functions each hard-code what their fallback is if they can't parse the value (`parseRetryAfterNumber` falls back to `parseRetryAfterString`, `parseRetryAfterString` falls back to a hard-coded 1 second).

This works for the small program we're making here, but if we were working on a program we wanted to grow, or to maintain over time, I'd expect "how do I do something" and "what do I do if it didn't work" to be treated separately (or "orthogonally" we'd often call this).

So I'd probably have each of these functions return either a `time.Duration` or an `error` (or alternatively, an optional `time.Duration` which may be `nil`, which in Go we do by returning a pointer), and then either have the caller, or some other function, select between them. So we'd have code somewhere (either in its own function, or in `handleStatusCode`) which reads something like:

```go
func parseRetryAfterHeader(retryAfter string) time.Duration {
    if duration := parseRetryAfterNumber(retryAfter); duration != nil {
        return *duration
    }
    if duration := parseRetryAfterString(retryAfter); duration != nil {
        return *duration
    }
    return time.Second
}
```

That way:
1. We can easily tell what the sequence of checks and fallbacks is in one place
2. If we wanted to re-use `parseRetryAfterString` somewhere where a different default made sense, or if we wanted to swap the order of which format we attempt first, we can easily do that.